balancer-js/src/modules/swaps/joinExit/actions/exit.ts

```diff
@@ -0,0 +1,141 @@
+  getActionMinOut,
+  getActionOutputRef,
+} from './helpers';
+
+/**
+ * Create a ExitAction with relevant info.
+ * @param swapType
+ * @param swap
+ * @param tokenInIndex
+ * @param tokenOutIndex
+ * @param opRefKey
+ * @param assets
+ * @param slippage
+ * @param user
+ * @param relayerAddress
+ * @returns
+ */
+export function createExitAction(
+  swap: SwapV2,
+  tokenInIndex: number,
+  tokenOutIndex: number,
+  opRefKey: number,
+  assets: string[],
+  slippage: string,
+  user: string,
+  relayerAddress: string
+): [ExitAction, number] {
+  const actionStep = getActionStep(
+    tokenInIndex,
+    tokenOutIndex,
+    swap.assetInIndex,
+    swap.assetOutIndex
+  );
+  // Will get actual amount if input or chain amount if part of chain
+  const amountIn = getActionAmount(
+    swap.amount,
+    ActionType.Exit,
+    actionStep,
+    opRefKey
+  );
+  // This will be 0 if not a mainTokenOut action otherwise amount using slippage
+  const minOut = getActionMinOut(swap.returnAmount ?? '0', slippage);
+  // This will set opRef for next chained action if required
+  const [opRef, newOpRefKey] = getActionOutputRef(
+    actionStep,
+    swap.assetOutIndex,
+    opRefKey
+  );
+  let sender = relayerAddress;
+  let hasTokenIn = false;
+  if (actionStep === ActionStep.Direct || actionStep === ActionStep.TokenIn) {
+    sender = user;
+    hasTokenIn = true;
+  }
+  // Send to relayer unless this is main token out
+  let hasTokenOut = false;
+  let toInternalBalance = true;
+  let receiver = relayerAddress;
+  if (actionStep === ActionStep.Direct || actionStep === ActionStep.TokenOut) {
+    receiver = user;
+    toInternalBalance = false;
+    hasTokenOut = true;
+  }
+
+  const exitAction: ExitAction = {
+    type: ActionType.Exit,
+    poolId: swap.poolId,
+    tokenOut: assets[swap.assetOutIndex],
+    bpt: assets[swap.assetInIndex],
+    opRef: opRef.key ? [opRef] : [],
+    minOut,
+    amountIn,
+    assets,
+    actionStep,
+    sender,
+    receiver,
+    toInternal: toInternalBalance,
+    hasTokenIn,
+    hasTokenOut,
+  };
+  return [exitAction, newOpRefKey];
+}
+
+/**
+ * Creates encoded exitPool call.
+ * @param pool
+ * @param action
+ * @param wrappedNativeAsset
+ * @returns
+ */
+export function buildExitCall(
+  pool: SubgraphPoolBase,
+  action: ExitAction,
+  wrappedNativeAsset: string
+): [string, string, string, ExitPoolData] {
+  const assets = pool.tokensList;
+  const assetHelpers = new AssetHelpers(wrappedNativeAsset);
+  // tokens must have same order as pool getTokens
+  const [sortedTokens] = assetHelpers.sortTokens(assets) as [string[]];
+  const exitToken = action.tokenOut;
+  const exitTokenIndex = sortedTokens.findIndex(
+    (t) => t.toLowerCase() === exitToken.toLowerCase()
+  );
+  const minAmountsOut = Array(assets.length).fill('0');
+  // Variable amount of token out (this has slippage applied)
+  minAmountsOut[exitTokenIndex] = action.minOut;
+  // Uses exact amount in
+  const bptAmtIn = action.amountIn;
```

Comment:
It's almost an irrelevant change, but I think bptAmtIn is a little out of the pattern of the other "amount" variable names, the other "amount" variable names aren't abbreviating "amount", I think bptAmountIn would be more befitting